Implement FieldMerge and FieldMergeFunction

dfalling opened the pull request:
I feel like this is pretty close. Something's wrong with the `@unboxed` though:
```ocaml
  open ApolloClient
  make(
    ~cache=Cache.InMemoryCache.make(
      ~typePolicies=[
        (
          "Trip",
          Cache.InMemoryCache.TypePolicy.make(
            ~fields=[
              (
                "elements",
                FieldPolicy({
                  merge: Some(MergeFunction(incomingMerge)),
                  read: None,
                  keyArgs: None,
                }),
              ),
            ],
            (),
          ),
        ),
```

Compiles to
```javascript
ApolloClient.make(undefined, undefined, undefined, Caml_option.some(httpLink), ApolloClient__Cache_InMemory_InMemoryCache.make(undefined, undefined, undefined, undefined, [
          [
            "Trip",
            ApolloClient__Cache_InMemory_Policies.TypePolicy.make([[
                    "elements",
                    {
                      TAG: /* FieldPolicy */3,
                      _0: {
                        keyArgs: undefined,
                        read: undefined,
                        merge: /* MergeFunction */{
                          _0: incomingMerge
                        }
                      }
                    }
                  ]], undefined, undefined, undefined, undefined, undefined)
          ]
```

How can I get rid of that extra `merge: { _0: incomingMerge }`?

dfalling commented on src/@apollo/client/cache/inmemory/ApolloClient__Cache_InMemory_Policies_FieldPolicy.res:
```diff
@@ -4,13 +4,6 @@ module FieldNode = ApolloClient__Graphql.Language.Ast.FieldNode
 module ReadFieldFunction = ApolloClient__Cache_Core_Types_Common.ReadFieldFunction
 module ToReferenceFunction = ApolloClient__Cache_Core_Types_Common.ToReferenceFunction
 
-module FieldMergeFunction = {
-  module Js_ = {
-    type t
-  }
-  type t = Js_.t
-}
-
```
I had to move this type further down in the module to share the `FieldFunctionOptions.t` type.

jeddeloh commented:
I'd have to look close at the typescript types, but on first glance this looks great. 

> How can I get rid of that extra merge: { _0: incomingMerge }?

As long as you're keeping with the style of the rest of the library you can't. This is the cost of (what I considered at the time) a better API.  So we can do stuff like `merge: MergeFunction(someFn)` vs. `merge: FieldMerge.mergeFunction(someFn)`. My opinion was that the configuration of this stuff was never going to be a performance issue and it was so much more ergonomic and clear to use a variant instead of hunting down constructor functions. In other words, if we dropped the variants entirely and went with the unboxed stuff directly we'd get what you expect.

dfalling commented:
Unfortunately that nesting breaks the feature. Apollo doesn't recognize the merge function and nothing happens. I need a way for the function to be assigned directly, without the _0.

jeddeloh commented:
I think the unboxed portion is working fine. The `_0` is unwrapped when passed to the `FieldMerge.toJs` [here](https://github.com/jeddeloh/rescript-apollo-client/pull/144/files#diff-e74fe96d773f2f2cbf50239d31b26258f1ee24bfa46c4cdb00cfd84028538d07R119).

I think that the actual issue is that [line 124](https://github.com/jeddeloh/rescript-apollo-client/pull/144/files#diff-e74fe96d773f2f2cbf50239d31b26258f1ee24bfa46c4cdb00cfd84028538d07R124) needs to be typed as a `Js_.t`
```diff
- let mergeFunction = (v: FieldMergeFunction.t<'existing>) => Any(v)
+ let mergeFunction = (v: FieldMergeFunction.Js_.t<'existing>) => Any(v)
```

That will then force you to call `FieldMergFunction.toJs` on [133](https://github.com/jeddeloh/rescript-apollo-client/pull/144/files#diff-e74fe96d773f2f2cbf50239d31b26258f1ee24bfa46c4cdb00cfd84028538d07R119)

dfalling commented:
> I think that the actual issue is that [line 124](https://github.com/jeddeloh/rescript-apollo-client/pull/144/files#diff-e74fe96d773f2f2cbf50239d31b26258f1ee24bfa46c4cdb00cfd84028538d07R124) needs to be typed as a Js_.t

Aah, ok. I fixed those references, but still am seeing the `_0`. I've cleaned the package and my entire codebase so I don't _think_ it's a caching issue.

One additional issue- the `true` isn't being output as `true`. It's output as `merge: /* True */0` which actually results in false, since 0 is falsey.

jeddeloh commented:
In the compiled file where you configured your client (what you showed previously), you should still be seeing `_0`. Also, `merge: /* True */0` is the correct representation of `FieldMerge.True` at that level. All of that is passed to `InMemoryCache.make` and then goes through a bunch of layers of conversion from the ReScript layer. If you jump through the compiled code it's a little obtuse, but it should eventually convert that `0` to `true` just before it get's passed to the actual Apollo function in javascript.

What have you tried in debugging the issue so far? Does a debugger in your merge function get called? You could also put some debuggers in`FieldMerge.toJs` for instance, to see what is actually going on there if you don't believe it's getting converted.

dfalling commented:
Oh got it- I missed the fact that the toJs step happens afterwards. Turns out my issue was with `yarn link` not picking up my changes. I ran it off my fork instead and it worked as expected. Thanks for the help working through this! Verified that `MergeFunction` and `True` work locally as expected.

jeddeloh commented:
Excellent! I really appreciate you putting in the PR. I'll merge and cut a release soon.
